thrust/testing/complex.cu

```diff
@@ -7,70 +8,231 @@
+      thrust::complex<T> construct_from_copy(expected);
+      ASSERT_EQUAL(expected.real(), construct_from_copy.real());
+      ASSERT_EQUAL(expected.imag(), construct_from_copy.imag());
+    }
+
+    {
+      thrust::complex<T> construct_from_move(thrust::complex<T>(real, imag));
+      ASSERT_EQUAL(real, construct_from_move.real());
+      ASSERT_EQUAL(imag, construct_from_move.imag());
+    }
+
+    {
+      thrust::complex<T> copy_assign{};
+      const thrust::complex<T> expected{real, imag};
+      copy_assign = expected;
+      ASSERT_EQUAL(expected.real(), copy_assign.real());
+      ASSERT_EQUAL(expected.imag(), copy_assign.imag());
+    }
+
+    {
+      thrust::complex<T> move_assign{};
+      const thrust::complex<T> expected{real, imag};
+      move_assign = thrust::complex<T>{real, imag};
+      ASSERT_EQUAL(expected.real(), move_assign.real());
+      ASSERT_EQUAL(expected.imag(), move_assign.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_lvalue_T{};
+      const thrust::complex<T> expected{real};
+      const T to_be_copied = real;
+      assign_from_lvalue_T = to_be_copied;
+      ASSERT_EQUAL(expected.real(), assign_from_lvalue_T.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_lvalue_T.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_rvalue_T{};
+      const thrust::complex<T> expected{real};
+      assign_from_rvalue_T = T(real);
+      ASSERT_EQUAL(expected.real(), assign_from_rvalue_T.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_rvalue_T.imag());
+    }
+
+    {
+      const std::complex<T> expected(real, imag);
+      const thrust::complex<T> copy_from_std(expected);
+      ASSERT_EQUAL(expected.real(), copy_from_std.real());
+      ASSERT_EQUAL(expected.imag(), copy_from_std.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_lvalue_std{};
+      const std::complex<T> expected(real, imag);
+      assign_from_lvalue_std = expected;
+      ASSERT_EQUAL(expected.real(), assign_from_lvalue_std.real());
+      ASSERT_EQUAL(expected.imag(), assign_from_lvalue_std.imag());
+    }
+
+    {
+      thrust::complex<T> assign_from_rvalue_std{};
+      assign_from_rvalue_std = std::complex<T>(real, imag);
+      ASSERT_EQUAL(real, assign_from_rvalue_std.real());
+      ASSERT_EQUAL(imag, assign_from_rvalue_std.imag());
+    }
   }
 };
-SimpleUnitTest<TestComplexConstructors, FloatingPointTypes> TestComplexConstructorsInstance;
+SimpleUnitTest<TestComplexConstructionAndAssignment, FloatingPointTypes>
+  TestComplexConstructionAndAssignmentInstance;
 
+template <typename T>
+struct TestComplexConstructionAndAssignmentWithPromoting
+{
+  void operator()(void)
+  {
+    using T0 = T;
+    typedef other_floating_point_type_t<T0>  T1;
+
+    thrust::host_vector<T1> data = unittest::random_samples<T1>(2);
+
+    const T0 real_T0 = data[0];
+    const T0 imag_T0 = data[1];
+
+    const T1 real_T1 = static_cast<T1>(real_T0);
+    const T1 imag_T1 = static_cast<T1>(imag_T0);
```

Comment:
This seems like some mixup.

The data vector is of type T1. Why are we converting to T0 and back to T1? Is the goal to adopt for precision loss? Otherwise I would do:
```suggestion
    thrust::host_vector<T0> data = unittest::random_samples<T0>(2);

    const T0 real_T0 = data[0];
    const T0 imag_T0 = data[1];

    const T1 real_T1 = static_cast<T1>(real_T0);
    const T1 imag_T1 = static_cast<T1>(imag_T0);
```

Reply:
Indeed, was just c&p error, should be

```cpp
thrust::host_vector<T0> data = unittest::random_samples<T0>(2);
```